### DISCUSSION TITLE
Preview deploys and databases

### DISCUSSION THREAD

**Discussion by osdiab** [vercel/next.js#22715] [Help]
It's become popular with modern cloud deployment services like Vercel, Netlify, Linc, and so on to deploy web apps on every commit for pull requests. This makes a lot of sense for frontend code.

It's also become popular, however, with frameworks like NextJS, to deploy one's API in the same codebase and infrastructure as their frontend code. But APIs often require a database to function, and databases often change schemas with migrations. To me, this means that preview deploys in a frontend/API monolith could often fail if the data model changes on a branch.

How have others handled this "Preview Deploys" development pattern, when databases get involved? Is there an elegant way to spin up separate database instances per preview deploy, that match the schemas/migrations defined per-branch, and work well with these very distributed, often serverless hosting providers?

Some initial thoughts

- is there a fast and cheap way to spin up a DB from a template on deploys (based on a project's main branch), run migrations of a particular branch on it, and have a Vercel preview deploy somehow discover that new DB instance? What if such a database has a lot of data, can it still be kept snappy? Maybe with Docker images?
- maybe a checked-in sqlite database might make sense to easily support preview deploys, so long as it's assumed that preview deploys don't experience much parallel access? but unless you're using sqlite as your prod DB, it creates a significant difference between preview and production environments
- Or perhaps coupling one's data model to preview deploys is a bad idea altogether and data access ought to be kept separate from this kind of frontend code.

Curious to hear what other people have done.

This is cross-posted to Stack Overflow: https://stackoverflow.com/questions/66455012/using-databases-with-deploys-on-each-commit

**Comment by mduecker** [comment]
There was an interesting Twitter thread on this, involving Guillermo Rauch and containing some points to potential solutions:
https://twitter.com/rauchg/status/1337094230875893760

**Comment by zomars** [comment]
Prisma Cloud [promised this](https://www.youtube.com/watch?v=oojjJh-51ZI) but hasn't delivered yet AFAIK

**Comment by peterp** [comment]
We've solved this at Snaplet with Preview Databases. You get the DB schema and production-like data "in an instant" for each pull-request preview. 

We even have a GitHub Action that co-ordinates all the work. Check it out: https://github.com/snaplet/vercel-action

**Reply by osdiab** [reply]
interested in this, but i think that your repo is private :)

**Reply by peterp** [reply]
Ooh, the link is wrong: https://github.com/snaplet/vercel-action
